File: Cloud Keys Samples/Amazon KMS/c++/amazonkms.cpp
/*
 * Cloud Keys 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Keys in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudkeys
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */



#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudkeys.h"

#define LINE_LEN 80

class MyAmazonKMS : public AmazonKMS {
public:
	virtual int FireKeyList(AmazonKMSKeyListEventParams* e) {
		printf("  %s\n", e->Id);
		return 0;
	}
};

void prompt(char* command, char* punctuation, char* default_str, char(&ret)[LINE_LEN]) {
	printf("%s [%s] %s ", command, default_str, punctuation);
	fgets(ret, LINE_LEN, stdin);
	if (strlen(ret) == 1) {
		strcpy(ret, default_str);
		return;
	}
	ret[strlen(ret) - 1] = '\0';
}

int main(int argc, char* argv[])
{
	int ret_code;

	if (argc < 1) {

		fprintf(stderr, "usage: amazonkms\n");
		fprintf(stderr, "\n");
		printf("\nPress enter to continue.\n");
		getchar();

	}
	else{
		MyAmazonKMS amazonkms; 
		char command[LINE_LEN];     // user's command
		char *argument;             // arguments to the user's command
		char keyname[LINE_LEN], keytype[LINE_LEN], algorithm[LINE_LEN];
		char *output, input[LINE_LEN];
		char* arn;
		int output_length;

		prompt("Access Key", ":", "", command);
		amazonkms.SetAccessKey(command);
		prompt("Secret Key", ":", "", command);
		amazonkms.SetSecretKey(command);

		while (1)
		{
			printf("\nAmazonKMS Commands:\n");
			printf("l            List Keys\n");
			printf("c            Create a Key\n");
			printf("d            Delete a Key\n");
			printf("e            Encrypts a String\n");
			printf("s            Signs a String\n");
			printf("?            Display Options\n");
			printf("q            Quit\n");

			printf("> ");
			fgets(command, LINE_LEN, stdin);
			command[strlen(command) - 1] = '\0';
			argument = strtok(command, " \t\n");

			if (!strcmp(command, "?"))
			{
				printf("\nAmazonKMS Commands:\n");
				printf("l            List Keys\n");
				printf("c            Create a Key\n");
				printf("d            Delete a Key\n");
				printf("e            Encrypts a String\n");
				printf("s            Signs a String\n");
				printf("?            Display Options\n");
				printf("q            Quit\n");

				printf("> ");
			}

			else if (!strcmp(command, "l"))
			{
				printf("Keys: \n");
				amazonkms.ListKeys();
				printf("\n");
			}

			else if (!strcmp(command, "c"))
			{
				prompt("Key Spec", ":", "RSA_2048", keyname);
				prompt("For signing", "?", "y", keytype);
				if (keytype[1] == 'y') {
					arn = amazonkms.CreateKey(keyname, true, "sign, verify");
				}
				else {
					arn = amazonkms.CreateKey(keyname, false, "encrypt, decrypt, sign, verify");
				}

				ret_code = amazonkms.GetLastErrorCode();
				if (ret_code) goto done;

				printf("\nKey Created Successfully\n");
				printf("Key ARN: %s\n",arn);
			}

			else if (!strcmp(command, "d"))
			{
				prompt("Key Id", ":", "", keyname);
				amazonkms.ScheduleKeyDeletion(keyname,7);

				ret_code = amazonkms.GetLastErrorCode();
				if (ret_code) goto done;

				printf("\nKey Deleted Successfully\n");
			}

			else if (!strcmp(command, "e"))
			{
				prompt("Key Id", ":", "", keyname);
				prompt("Encryption Algorithm", ":", "RSAES_OAEP_SHA_256", algorithm);
				prompt("String to Encrypt", ":", "This is an Example!", input);
				amazonkms.SetInputData(input, sizeof(input));
				amazonkms.Encrypt(keyname, algorithm);

				printf("Encrypted Data: ");
				amazonkms.GetOutputData(output, output_length);
				printf(output);
			}

			else if (!strcmp(command, "s"))
			{
				prompt("Key Id", ":", "", keyname);
				prompt("Signature Algorithm", ":", "ECDSA_SHA_256", algorithm);
				prompt("String to Sign", ":", "This is an Example!", input);
				amazonkms.SetInputData(input, sizeof(input));
				amazonkms.Sign(keyname, algorithm, false);
				printf("Signature Data: ");
				amazonkms.GetOutputData(output, output_length);
				printf(output);
			}

			else if (!strcmp(command, "q")) 
			{
				goto done;
			}

			else
			{
				printf("Command Not Recognized.\n");
			}

		  ret_code = amazonkms.GetLastErrorCode();
			if (ret_code) goto done;

		}  // end of main while loop

	done:
		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (amazonkms.GetLastError())
			{
				printf(" \"%s\"\n", amazonkms.GetLastError());
			}
		}
		fprintf(stderr, "\npress <return> to continue...\n");
		exit(ret_code);
		return 0;
	}
}

File: Cloud Keys Samples/Amazon KMS/netcore/amazonkms.cs
/*
 * Cloud Keys 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Keys in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudkeys
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.CloudKeys;

class amazonkmsDemo
{
  private static AmazonKMS amazonkms = new nsoftware.CloudKeys.AmazonKMS();
  private static string keyId = "";

  static void Main(string[] args)
  {
    try
    {
      amazonkms.OnKeyList += amazonkms_OnKeyList;
      amazonkms.OnError += amazonkms_OnError;

      // Prompt for authentication information.
      Console.Write("Enter your access key: ");
      amazonkms.AccessKey = Console.ReadLine();
      Console.Write("Enter your secret key: ");
      amazonkms.SecretKey = Console.ReadLine();

      // Process user commands.
      Console.WriteLine("Type \"?\" or \"help\" for a list of commands.");
      Console.Write("amazonkms> ");
      string command;
      string[] arguments;

      while (true)
      {
        command = Console.ReadLine();
        arguments = command.Split();

        if (arguments[0] == "?" || arguments[0] == "help")
        {
          Console.WriteLine("Commands: ");
          Console.WriteLine("  ?                                 display the list of valid commands");
          Console.WriteLine("  help                              display the list of valid commands");
          Console.WriteLine("  ls                                list keys");
          Console.WriteLine("  create <type> <forsigning>        create a new key");
          Console.WriteLine("    ex. create RSA_2048 y");
          Console.WriteLine("    ex. create SYMMETRIC_DEFAULT n");
          Console.WriteLine("  del <id> <days>                   schedule the key with the specified id to be deleted in the specified number of days");
          Console.WriteLine("  encrypt <id> <alg> <string>       encrypt a string");
          Console.WriteLine("    ex. encrypt 1234abcd-12ab-34cd-56ef-1234567890ab SYMMETRIC_DEFAULT test");
          Console.WriteLine("  sign <id> <alg> <string>          sign a string");
          Console.WriteLine("    ex. sign 1234abcd-12ab-34cd-56ef-1234567890ab RSASSA_PSS_SHA_256 test");
          Console.WriteLine("  quit                              exit the application");
        }
        else if (arguments[0] == "ls")
        {
          amazonkms.ListKeys();
        }
        else if (arguments[0] == "create")
        {
          if (arguments.Length > 2)
          {
            bool forSigning = arguments[2].ToLower() == "y";
            keyId = amazonkms.CreateKey(arguments[1], forSigning, "sign, verify");
            Console.WriteLine("Key created successfully.");
            Console.WriteLine("New key id: " + keyId);
          }
        }
        else if (arguments[0] == "del")
        {
          if (arguments.Length > 2)
          {
            amazonkms.ScheduleKeyDeletion(arguments[1], int.Parse(arguments[2]));
            Console.WriteLine("Key scheduled for deletion.");
          }
        }
        else if (arguments[0] == "encrypt")
        {
          if (arguments.Length > 3)
          {
            string text = "";
            for (int i = 3; i < arguments.Length; i++)
            {
              text += arguments[i] + " ";
            }
            amazonkms.InputData = text;

            amazonkms.Encrypt(arguments[1], arguments[2]);
            Console.WriteLine("Encrypted data: ");
            Console.WriteLine(amazonkms.OutputData);
          }
        }
        else if (arguments[0] == "sign")
        {
          if (arguments.Length > 3)
          {
            string text = "";
            for (int i = 3; i < arguments.Length; i++)
            {
              text += arguments[i] + " ";
            }
            amazonkms.InputData = text;

            amazonkms.Sign(arguments[1], arguments[2], false);
            Console.WriteLine("Signature data: ");
            Console.WriteLine(amazonkms.OutputData);
          }
        }
        else if (arguments[0] == "quit")
        {
          break;
        }
        else if (arguments[0] == "")
        {
          // Do nothing.
        }
        else
        {
          Console.WriteLine("Invalid command.");
        } // End of command checking.

        Console.Write("amazonkms> ");
      }
    }
    catch (Exception ex)
    {
      Console.WriteLine(ex.Message);
    }
  }

  private static void amazonkms_OnKeyList(object sender, AmazonKMSKeyListEventArgs e)
  {
    Console.WriteLine("   " + e.ARN);
  }

  private static void amazonkms_OnError(object sender, AmazonKMSErrorEventArgs e)
  {
    Console.WriteLine("Error [" + e.ErrorCode + "]: " + e.Description);
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add an key to the dictionary for each argument
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/" then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}